src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceProviderServiceExtensions.cs

```diff
@@ -69,6 +71,7 @@ public static object GetRequiredService(this IServiceProvider provider, Type ser
         /// <param name="provider">The <see cref="IServiceProvider"/> to retrieve the service object from.</param>
         /// <returns>A service object of type <typeparamref name="T"/>.</returns>
         /// <exception cref="System.InvalidOperationException">There is no service of type <typeparamref name="T"/>.</exception>
+        [return: NotNull]
```

Comment:
Would it be better to constrain T to be non-nullable?
```C#
public static T GetRequiredService<T>(this IServiceProvider provider) where T : notnull
```
?

Reply:
(Other APIs, like the one below, could use that constraint as well.)

Reply:
@davidfowl and I went over some of the other APIs here. We added the `notnull` constraint to another type, but the rest of these APIs don't qualify for additional constraints.